sike/sikep751/sikep751.go

```diff
@@ -0,0 +1,168 @@
+type SIKESecretKey struct {
+	Scalar    [MessageSize]byte
+	SecretKey *SIDHSecretKeyBob
+	PublicKey *SIDHPublicKeyBob
+}
+
+// SIKEPublicKey is just a wrapper around SIDHPublicKeyBob
+type SIKEPublicKey struct {
+	PublicKey *SIDHPublicKeyBob
+}
+
+// SIKECipherText contains SIDHSecretKeyBob, SIDHPublicKeyBob, and Scalar
+type SIKECipherText struct {
+	PublicKey *SIDHPublicKeyAlice
+	Scalar    [MessageSize]byte
+}
+
+// SIKESharedSecret contains SIDHSecretKeyBob, SIDHPublicKeyBob, and Scalar
+type SIKESharedSecret struct {
+	Scalar [SharedSecretSize]byte
+}
+
+// GenerateKeyPair generation generates SIKE secret-key and public-key.
+// The secret-key contains a random message + secret-key + public-key.
+func GenerateKeyPair(rand io.Reader) (publicKey *SIKEPublicKey, sikeSecretKey *SIKESecretKey, err error) {
+
+	publicKey = new(SIKEPublicKey)
+	sikeSecretKey = new(SIKESecretKey)
+	var secretKey = new(SIDHSecretKeyBob)
+
+	// Randomly generate 32 byte s
+	_, err = io.ReadFull(rand, sikeSecretKey.Scalar[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Generate Encryptor secret-key and public-key
+	publicKey.PublicKey, secretKey, err = GenerateBobKeypair(rand)
+
+	// Copy secretKey and publicKey into SIKEsecretKey
+	sikeSecretKey.SecretKey = secretKey
+	sikeSecretKey.PublicKey = publicKey.PublicKey
+
+	return
+}
+
+// Encapsulation gets the public-key as input and generates
+// the SIKE ciphertex and shared secret. The generated ciphertet contains
+// the public-key and a random message
+func Encapsulation(rand io.Reader, publicKey *SIKEPublicKey) (cipherText *SIKECipherText, sharedSecret *SIKESharedSecret, err error) {
+	cipherText = new(SIKECipherText)
+	sharedSecret = new(SIKESharedSecret)
+	var ephemeralSk = new(SIDHSecretKeyAlice)
+	var jinvariant [SharedSecretSize]byte
+	var h [MessageSize]byte
+	var tmp = make([]byte, (CiphertextSize + MessageSize))
+
+	// Generate ephemeral secretKey G(m||pk) mod oA
+	_, err = io.ReadFull(rand, tmp[:MessageSize])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Append publicKey to message and hash it
+	publicKey.PublicKey.ToBytes(tmp[MessageSize:])
+	CShakeSum256(ephemeralSk.Scalar[:], tmp[:CiphertextSize], []byte(G))
+
+	// Perform mod oA
+	ephemeralSk.Scalar[SecretKeySize-1] = MaskAliceByte1
+	ephemeralSk.Scalar[SecretKeySize-2] &= MaskAliceByte2 // clear high bits, so scalar < 2^372
+	ephemeralSk.Scalar[0] &= MaskAliceByte3               // clear low bit, so scalar is even
+
+	// Encryption
+	var tmpPk = new(SIDHPublicKeyAlice)
+	*tmpPk = ephemeralSk.PublicKey()
+	cipherText.PublicKey = tmpPk
+	jinvariant = ephemeralSk.SharedSecret(publicKey.PublicKey)
+	CShakeSum256(h[:], jinvariant[:], []byte(P))
+
+	for i := 0; i < MessageSize; i++ {
+		cipherText.Scalar[i] = tmp[i] ^ h[i]
+	}
+
+	// Generate shared secret: ss = H(m||ct)
+	cipherText.PublicKey.ToBytes(tmp[MessageSize:])
+	copy(tmp[CiphertextSize:], cipherText.Scalar[:])
```

Comment:
I think that's a bug.
1. Public key is serialized to &tmp[32] and it wrties 596 bytes
2. Then scalar is coppied to the location &tmp[596] an 32 bytes are written. Overwritting 1.

Line 123 and 124 seem are not needed, it should be:
```
copy(tmp[MessageSize:], cipherText.Scalar[:])
```